source/common/stats/symbol_table_impl.h

```diff
@@ -0,0 +1,136 @@
+ * the same underlying symbol.
+ */
+class SymbolTableImpl : public SymbolTable {
+public:
+  StatNamePtr encode(const absl::string_view name) override;
+
+  // For testing purposes only.
+  size_t size() const override {
+    RELEASE_ASSERT(encode_map_.size() == decode_map_.size(), "");
+    return encode_map_.size();
+  }
+
+private:
+  friend class StatNameImpl;
+  friend class StatNameTest;
+
+  struct SharedSymbol {
+    Symbol symbol_;
+    uint32_t ref_count_;
+  };
+
+  /**
+   * Decodes a vector of symbols back into its period-delimited stat name.
+   * If decoding fails on any part of the symbol_vec, we release_assert and crash hard, since this
+   * should never happen, and we don't want to continue running with a corrupt stats set.
+   *
+   * @param symbol_vec the vector of symbols to decode.
+   * @return std::string the retrieved stat name.
+   */
+  std::string decode(const SymbolVec& symbol_vec) const;
+
+  /**
+   * Since SymbolTableImpl does manual reference counting, a client of SymbolTable (such as
+   * StatName) must manually call free(symbol_vec) when it is freeing the stat it represents. This
+   * way, the symbol table will grow and shrink dynamically, instead of being write-only.
+   *
+   * @param symbol_vec the vector of symbols to be freed.
+   */
+  void free(const SymbolVec& symbol_vec);
+
+  /**
+   * Convenience function for encode(), symbolizing one string segment at a time.
+   *
+   * @param sv the individual string to be encoded as a symbol.
+   * @return Symbol the encoded string.
+   */
+  Symbol toSymbol(absl::string_view sv);
+
+  /**
+   * Convenience function for decode(), decoding one symbol at a time.
+   *
+   * @param symbol the individual symbol to be decoded.
+   * @return absl::string_view the decoded string.
+   */
+  absl::string_view fromSymbol(const Symbol symbol) const;
+
+  // Stages a new symbol for use. To be called after a successful insertion.
+  void newSymbol() {
+    if (pool_.empty()) {
+      RELEASE_ASSERT(monotonic_counter_ < std::numeric_limits<Symbol>::max(), "");
+      current_symbol_ = ++monotonic_counter_;
+    } else {
+      current_symbol_ = pool_.top();
+      pool_.pop();
+    }
+  }
+
+  Symbol monotonicCounter() { return monotonic_counter_; }
+
+  // Stores the symbol to be used at next insertion. This should exist ahead of insertion time so
+  // that if insertion succeeds, the value written is the correct one.
+  Symbol current_symbol_ = 0;
+
+  // If the free pool is exhausted, we monotonically increase this counter.
+  Symbol monotonic_counter_ = 0;
+
+  // Bimap implementation.
+  // The encode map stores both the symbol and the ref count of that symbol.
+  // Using absl::string_view lets us only store the complete string once, in the decode map.
+  std::unordered_map<absl::string_view, SharedSymbol, StringViewHash> encode_map_;
+  std::unordered_map<Symbol, std::string> decode_map_;
+
+  // Free pool of symbols for re-use.
+  std::stack<Symbol> pool_;
```

Comment:
You might be able to optimize the pool_ -- depending on usage pattern, so that if you free all the highest-numbered symbols, you remove them from the pool_ and adjust monotonic_counter_ (thus making it non-monotonic I suppose). But this is probably fairly minor. I don't actually expect you are going to wind up freeing huge numbers of symbols in practice, based on how the stats are constructed.

Reply:
I think the issue with decrementing the monotonic counter is that increasing it again could cause a collision -- I would have to find a clever way to avoid that. I'll put a TODO for myself here to think about this.